Dewhitelisting contracts

denalimarsh opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: https://github.com/cosmos/peggy/issues/140

This PR implements the contract portion of the dewhitelisting/namespacing solution and includes changes from https://github.com/cosmos/peggy/pull/182, https://github.com/cosmos/peggy/pull/184.

The following snippet contains the exact cmds that were used to end-to-end test this feature:
```
---------------------------------------------- Setup ---------------------------------------------
ebd unsafe-reset-all
ebd start
yarn develop
yarn peggy:all
ebrelayer generate
ebrelayer init tcp://localhost:26657 ws://localhost:7545/ 0x30753E4A8aad7F8597332E813735Def5dD395028 validator --chain-id=peggy

--------------------------- Native Ethereum assets to Cosmos and back ---------------------------
# Lock TEST token or Ethereum on contracts
# yarn peggy:lock [cosmos-recipient-address] [token-contract-address] [amount]
yarn peggy:lock $(ebcli keys show testuser -a) 0x345cA3e014Aaf5dcA488057592ee47305D9B3e10 10
yarn peggy:lock $(ebcli keys show testuser -a) eth 10

# Enter keyring password to send tx, relayer will log the transaction hash
# Query tx by txhash
ebcli q tx [tx-hash]
# In the logs you should see an event type "prophecy_status" with attribute "status: success"
# If you see "prophecy already finalized: failed to execute message; message index: 0" then reset your bridge chain with `ebd unsafe-reset-all`

 # Query cosmos recipient, should have a 'peggytest' or 'peggyeth' balance of 10
 ebcli q auth account $(ebcli keys show testuser -a)

# Burn 1 'peggytest' or 'peggyeth' coin with Ethereum recipient accounts[9]
# ebcli tx ethbridge burn [cosmos-sender-address] [ethereum-receiver-address] [amount] [symbol] --ethereum-chain-id [ethereum-chain-id] [flags]
ebcli tx ethbridge burn $(ebcli keys show testuser -a) 0x5AEDA56215b167893e80B4fE645BA6d5Bab767DE 1 peggytest --ethereum-chain-id=5777 --from=testuser --yes
ebcli tx ethbridge burn $(ebcli keys show testuser -a) 0x5AEDA56215b167893e80B4fE645BA6d5Bab767DE 1 peggyeth --ethereum-chain-id=5777 --from=testuser --yes

# Query ethereum recipient with `yarn peggy:getTokenBalance [ethereum-address] [token-contract-address]`
# ERC20: balance should be 0.000000000000000001 TEST (1 TEST with 18 decimals)
yarn peggy:getTokenBalance 0x5AEDA56215b167893e80B4fE645BA6d5Bab767DE 0x345cA3e014Aaf5dcA488057592ee47305D9B3e10
# ETH: balance should be increased by 1 Wei to 100.000000000000000001 Eth (100000000000000000001 Wei)
yarn peggy:getTokenBalance 0x5AEDA56215b167893e80B4fE645BA6d5Bab767DE eth

-------------------------------- Native Cosmos assets to Ethereum -------------------------------
# Send 10 stake tokens to testuser
ebcli tx send validator $(ebcli keys show testuser -a) 10stake --yes

 # Query the account, should have a 'stake' balance of 10
ebcli q account $(ebcli keys show testuser -a)

# ebcli tx ethbridge lock [cosmos-sender-address] [ethereum-receiver-address] [amount] --ethereum-chain-id [ethereum-chain-id] [flags]
ebcli tx ethbridge lock $(ebcli keys show testuser -a) 0x5AEDA56215b167893e80B4fE645BA6d5Bab767DE 1 stake --ethereum-chain-id=5777 --from=testuser --yes

# Query account using the PEGGYSTAKE token contract address logged by the relayer

# yarn peggy:getTokenBalance [ethereum-address] [token-contract-address]
# Should have balance of 0.000000000000000001 PEGGYSTAKE (1 PEGGYSTAKE with 18 decimals)
yarn peggy:getTokenBalance 0x5AEDA56215b167893e80B4fE645BA6d5Bab767DE 0x409Ba3dd291bb5D48D5B4404F5EFa207441F6CbA
```

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [x] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [x] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

denalimarsh commented:
Changing the base branch from `master` to `dewhitelisting_modules`. Note to reviewers - this will show the files from https://github.com/cosmos/peggy/pull/183 as changes.

musnit commented on cmd/ebrelayer/contract/generated/bin/Valset/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
can we replace these all with rules in our base .gitignore file??

denalimarsh replied:
probably yeah - want to ignore the .bin and .abi files but not the directories, if I recall the generator throws an err when it can't find the proper directory to write to. Could also update the generator instead if global .gitignore rules fall short.

musnit commented on cmd/ebrelayer/types/types.go:
```diff
@@ -147,13 +147,15 @@ const (
 // String returns the event type as a string
 func (d CosmosMsgAttributeKey) String() string {
-	return [...]string{"unsupported", "cosmos_sender", "ethereum_receiver", "amount", "token_contract_address"}[d]
+	return [...]string{"unsupported", "cosmos_sender", "ethereum_receiver", "amount", "symbol", "token_contract_address"}[d]
```
you can remove all these types and just use the AttributeKey types included from EthBridge types if u want

musnit commented on testnet-contracts/contracts/CosmosBridge.sol:
```diff
@@ -149,22 +149,37 @@ contract CosmosBridge {
+                "Not enough locked Ethereum assets to complete the proposed prophecy"
+            );
             claimType = ClaimType.Burn;
+            tokenAddress = _tokenAddress;
```
_tokenAddress is no longer available in burn/lock relays, so I think we need to use bridgeBank.getControlledBridgeToken for the burn too. can discuss on slack if needed.

denalimarsh replied:
Removed `_tokenAddress` in latest.

okwme commented on cmd/ebrelayer/contract/generated/bindings/bridgeregistry/BridgeRegistry.go:
```diff
@@ -31,7 +31,7 @@ var (
 
 // BridgeRegistryBin is the compiled bytecode used for deploying new contracts.
-const BridgeRegistryBin = `608060405234801561001057600080fd5b506040516080806105378339810180604052608081101561003057600080fd5b8101908080519060200190929190805190602001909291908051906020019092919080519060200190929190505050836000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555082600160006101000a81548173ffffffffffffffffffffffffff ...
+const BridgeRegistryBin = `608060405234801561001057600080fd5b506040516080806105378339810180604052608081101561003057600080fd5b8101908080519060200190929190805190602001909291908051906020019092919080519060200190929190505050836000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555082600160006101000a81548173ffffffffffffffffffffffffff ...
```
should generated files be stored at all?

denalimarsh replied:
ebrelayer build will fail circleci without the contract bindings.

denalimarsh replied:
can see in circleci for https://github.com/cosmos/peggy/pull/185/commits/0daf985266b23261f7c096d4a260ef11f60925e5 (failed) and https://github.com/cosmos/peggy/pull/185/commits/e42d614b05c8ce1a8835eda7eff9746e93be6ba3 (succeeded)

okwme commented:
> Changing the base branch from `master` to `dewhitelisting_modules`. Note to reviewers - this will show the files from #183 as changes.

Since #183 has been merged into master should this be based against master as well now?

okwme commented:
Going through instructions from scratch i get the following error:
```bash
> ebrelayer generate

ERROR: exit status 127
```
on step 2 [here](https://github.com/denalimarsh/peggy/blob/dm-namespacing-solidity/docs/setup-relayer.md)

okwme commented:
I had to merge my changes from the contract test and might have merged conflicts incorrectly. I can try again after that PR has been merged into this branch, or if you're able to reproduce the error maybe it's something else.

denalimarsh commented:
> Since #183 has been merged into master should this be based against master as well now?

It's based against `cosmos:dewhitelisting_modules` so changes from #182 aren't duplicated here. See https://github.com/cosmos/peggy/pull/182#issuecomment-614698552. If it's easier to review the entire thing at once then yes can base against master.

denalimarsh commented:
> I had to merge my changes from the contract test and might have merged conflicts incorrectly. I can try again after that PR has been merged into this branch, or if you're able to reproduce the error maybe it's something else.

After merging in the updates from https://github.com/cosmos/peggy/pull/184 I get the same issue. The problem is that solc has specific requirements for importing/linking contracts and the cmds in `ebrelayer/contract/generator.go` are no longer compatible with the import updates from https://github.com/cosmos/peggy/pull/185/commits/27bd619f02832676448c1acdfd1fef3832f6e9ca.

okwme commented:
```bash
Error: Cannot find module 'big-integer'
```
Fixed w `yarn add big-integer`

okwme commented:
ah I see what the problem was with the imports. When you install with `yarn` the `node_modules` are primarily stored in the parent directory. When you install with `npm i` they are stored in the `testnet-contracts` directory. That made the imports break when they were explicit. Do you know a way to make `npm i` and `yarn` behave the same?

denalimarsh commented:
> ah I see what the problem was with the imports. When you install with `yarn` the `node_modules` are primarily stored in the parent directory. When you install with `npm i` they are stored in the `testnet-contracts` directory. That made the imports break when they were explicit. Do you know a way to make `npm i` and `yarn` behave the same?

Not off the top of my head, but it seems easy. We can also just change the readme to say only yarn is supported as all the scripts use it.

denalimarsh commented:
@okwme I'm the middle of debugging an issue on CosmosBridge.sol contract, hence the changes in https://github.com/cosmos/peggy/pull/185/commits/75fed97147804c2de49873033ef1bcb3d70cd0f3. Added the WIP label until I've pushed a fix.

denalimarsh commented:
@okwme pushed a fix and removed the blocked tag, this PR is R4R. I've included the cmds I used to end-to-end test this feature in the PR's description for reference.

musnit commented:
@denalimarsh can you rebase onto master/remove conflicts?

denalimarsh commented:
> @denalimarsh can you rebase onto master/remove conflicts?

This PR's base branch has been updated to master and I'm not seeing any conflicts.

denalimarsh commented:
@musnit are there additional changes in https://github.com/cosmos/peggy/pull/182 not reflected in this PR?

okwme commented:
I'm still getting the `ebrelayer generate` error:
```
> ebrelayer generate

ERROR: exit status 127
```
I used yarn and all open-zeppelin dependencies are installed in the root `node_modules`. All the contracts reference the dependencies with explicit location.

For contracts in `/test_contracts/contracts/` like:
```
import "../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
```
And for contracts in `test_contracts/contracts/BridgeBank/` like:
```
import "../../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
```

Ran instructions from a fresh install, any idea why this would be happening?

okwme commented:
Looks like it might be because i don't have `solc` installed globally? What's the best way to include that dependency?

after installing solc (v5) globally i'm still getting errors. I logged out the command throwing them within the compile and saw it failing on:
```
		err := execCmd(contractBINGenCmd)
```
That command `contractBINGenCmd` logs out to be:
```
solc --bin ./testnet-contracts/contracts/BridgeBank/BridgeBank.sol -o ./cmd/ebrelayer/contract/generated/bin/BridgeBank --overwrite --allow-paths *,
```
When I run that manually I get this:
```
zsh: no matches found: *,
```
When I remove the trailing comma I get this:
```
"app" is not a valid file.
```
When I remove the `--alow-paths` I get: 
```
./testnet-contracts/contracts/BridgeBank/BridgeBank.sol:5:1: Error: Source "./testnet-contracts/contracts/Oracle.sol" not found: File outside of allowed directories.
import "../Oracle.sol";
^---------------------^
./testnet-contracts/contracts/BridgeBank/BridgeBank.sol:6:1: Error: Source "./testnet-contracts/contracts/CosmosBridge.sol" not found: File outside of allowed directories.
import "../CosmosBridge.sol";
^---------------------------^
./testnet-contracts/contracts/BridgeBank/CosmosBank.sol:3:1: Error: Source "./node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol" not found: File outside of allowed directories.
import "../../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
^-------------------------------------------------------------------------------^
./testnet-contracts/contracts/BridgeBank/EthereumBank.sol:3:1: Error: Source "./node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol" not found: File outside of allowed directories.
import "../../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
^-------------------------------------------------------------------------------^
./testnet-contracts/contracts/BridgeBank/BridgeToken.sol:3:1: Error: Source "./node_modules/openzeppelin-solidity/contracts/token/ERC20/ERC20Mintable.sol" not found: File outside of allowed directories.
import "../../../node_modules/openzeppelin-solidity/contracts/token/ERC20/ERC20Mintable.sol";
^-------------------------------------------------------------------------------------------^
./testnet-contracts/contracts/BridgeBank/BridgeToken.sol:4:1: Error: Source "./node_modules/openzeppelin-solidity/contracts/token/ERC20/ERC20Detailed.sol" not found: File outside of allowed directories.
import "../../../node_modules/openzeppelin-solidity/contracts/token/ERC20/ERC20Detailed.sol";
^-------------------------------------------------------------------------------------------^
```

denalimarsh commented:
> When I run that manually I get this:
> 
> ```
> zsh: no matches found: *,
> ```

@okwme does the cmd work if you wrap the * and/or the comma in quotes i.e.
```
solc --bin ./testnet-contracts/contracts/BridgeBank/BridgeBank.sol -o ./cmd/ebrelayer/contract/generated/bin/BridgeBank --overwrite --allow-paths "*",
solc --bin ./testnet-contracts/contracts/BridgeBank/BridgeBank.sol -o ./cmd/ebrelayer/contract/generated/bin/BridgeBank --overwrite --allow-paths "*,"
```

okwme commented:
ok i see the problem is actually now coming from the `GenerateBindings` command. That's because it expects `abigen` to be globally available on the machine? Does that mean geth has to be installed globally?

Is it correct that for this to run you have to manually install `solc` and `abigen`? They're references as executable commands but i don't see them listed in the modules file. Do these need to be added there, somewhere else or just added as explicit instructions or requirements for the software to run?

okwme commented:
Also seeing you can't use node v14 because of a bug in ganache https://github.com/trufflesuite/ganache-cli/issues/732

okwme commented on docs/cosmos-to-ethereum.md:
```diff
@@ -109,11 +87,11 @@ yarn peggy:getTokenBalance 0x11111111262b236c9ac9a9a8c8e4276b5cf6b2c9 eth
-# ebcli tx ethbridge burn [cosmos-sender-address] [ethereum-receiver-address] [amount] --ethereum-chain-id [ethereum-chain-id] --token-contract-address [token-contract-address] [flags]
-ebcli tx ethbridge burn $(ebcli keys show testuser -a) 0x11111111262b236c9ac9a9a8c8e4276b5cf6b2c9 1eth  --ethereum-chain-id 3 --token-contract-address 0x0000000000000000000000000000000000000000 --from testuser --yes
+# ebcli tx ethbridge burn [cosmos-sender-address] [ethereum-receiver-address] [amount] --ethereum-chain-id [ethereum-chain-id [flags]
+ebcli tx ethbridge burn $(ebcli keys show testuser -a) 0x11111111262b236c9ac9a9a8c8e4276b5cf6b2c9 1eth  --ethereum-chain-id 3 --from testuser --yes
```
```suggestion
ebcli tx ethbridge burn $(ebcli keys show testuser -a) 0x11111111262b236c9ac9a9a8c8e4276b5cf6b2c9 1 peggyeth  --ethereum-chain-id 3 --from testuser --yes
```

denalimarsh commented:
> Is it correct that for this to run you have to manually install `solc` and `abigen`?

Ah, I've got them globally installed. I added them to the docs in https://github.com/cosmos/peggy/pull/185/commits/2dd6b588c43c1275184d24036817aca04da3c77c, if the resources provided are insufficient or confusing let me know.

> Also seeing you can't use node v14 because of a bug in ganache [trufflesuite/ganache-cli#732](https://github.com/trufflesuite/ganache-cli/issues/732)

Good find. I'm currently using node v10.16.3. Added a note about the issue in https://github.com/cosmos/peggy/pull/185/commits/e6bb9c43ea8f797bd691c729e6fdb3b02702a010.

denalimarsh commented:
@okwme I noticed that adding code coverage caused [a recent](https://circleci.com/gh/cosmos/peggy/277
) CircleCI job to fail, but it seems to be working again now. Anything to be concerned about?

okwme commented:
Thanks for final updates I think this is ready to merge 🎉
@marbar3778 was checking the CircleCI error, did you find anything?
